src/main/java/net/neoforged/neoforge/client/model/generators/template/ExtendedModelTemplateBuilder.java

```diff
@@ -0,0 +1,178 @@
+            ext.elements.forEach(elem -> builder.elements.add(elem.copy()));
+            builder.customLoader = ext.customLoader != null ? ext.customLoader.copy() : null;
+            builder.rootTransforms.copyFrom(ext.rootTransforms);
+            builder.renderType = ext.renderType;
+            builder.ambientOcclusion = ext.ambientOcclusion;
+            builder.guiLight = ext.guiLight;
+        }
+        return builder;
+    }
+
+    public static ExtendedModelTemplateBuilder builder() {
+        return new ExtendedModelTemplateBuilder();
+    }
+
+    public ExtendedModelTemplateBuilder parent(ResourceLocation parent) {
+        this.parent = Optional.of(parent);
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder suffix(String suffix) {
+        this.suffix = Optional.of(suffix);
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder requiredTextureSlot(TextureSlot slot) {
+        this.requiredSlots.add(slot);
+        return this;
+    }
+
+    /**
+     * Set the render type for this model.
+     *
+     * @param renderType the render type. Must be registered via
+     *                   {@link RegisterNamedRenderTypesEvent}
+     * @return this builder
+     * @throws NullPointerException if {@code renderType} is {@code null}
+     */
+    public ExtendedModelTemplateBuilder renderType(String renderType) {
+        Preconditions.checkNotNull(renderType, "Render type must not be null");
+        return renderType(ResourceLocation.parse(renderType));
+    }
+
+    /**
+     * Set the render type for this model.
+     *
+     * @param renderType the render type. Must be registered via
+     *                   {@link RegisterNamedRenderTypesEvent}
+     * @return this builder
+     * @throws NullPointerException if {@code renderType} is {@code null}
+     */
+    public ExtendedModelTemplateBuilder renderType(ResourceLocation renderType) {
+        Preconditions.checkNotNull(renderType, "Render type must not be null");
+        this.renderType = renderType;
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder transforms(Consumer<TransformsBuilder> action) {
+        action.accept(transforms);
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder ao(boolean ao) {
+        this.ambientOcclusion = ao;
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder guiLight(UnbakedModel.GuiLight light) {
+        this.guiLight = light;
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder element(Consumer<ElementBuilder> action) {
+        Preconditions.checkState(
+                customLoader == null || customLoader.allowInlineElements,
+                "Custom model loader %s does not support inline elements",
+                customLoader != null ? customLoader.loaderId : null);
+        ElementBuilder ret = new ElementBuilder();
+        action.accept(ret);
+        elements.add(ret);
+        return this;
+    }
```

Comment:
These methods need some short doc comments and the `ao()` method should be renamed to `ambientOcclusion()` for clarity.